container/containerd/handler.go

```diff
@@ -0,0 +1,244 @@
+
+// Handler for containerd containers.
+package containerd
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	libcontainerconfigs "github.com/opencontainers/runc/libcontainer/configs"
+	"golang.org/x/net/context"
+
+	"github.com/google/cadvisor/container"
+	"github.com/google/cadvisor/container/common"
+	containerlibcontainer "github.com/google/cadvisor/container/libcontainer"
+	"github.com/google/cadvisor/fs"
+	info "github.com/google/cadvisor/info/v1"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+type containerdContainerHandler struct {
+	client             containerdClient
+	name               string
+	id                 string
+	aliases            []string
+	machineInfoFactory info.MachineInfoFactory
+	// Absolute path to the cgroup hierarchies of this container.
+	// (e.g.: "cpu" -> "/sys/fs/cgroup/cpu/test")
+	cgroupPaths map[string]string
+	// Manager of this container's cgroups.
+	cgroupManager cgroups.Manager
+	fsInfo        fs.FsInfo
+	poolName      string
+	// Time at which this container was created.
+	creationTime time.Time
+	// Metadata associated with the container.
+	labels map[string]string
+	envs   map[string]string
+	// The container PID used to switch namespaces as required
+	pid int
+	// Image name used for this container.
+	image string
+	// The host root FS to read
+	rootFs string
+	// Filesystem handler.
+	ignoreMetrics container.MetricSet
+}
+
+var _ container.ContainerHandler = &containerdContainerHandler{}
+
+// newContainerdContainerHandler returns a new container.ContainerHandler
+func newContainerdContainerHandler(
+	client containerdClient,
+	name string,
+	machineInfoFactory info.MachineInfoFactory,
+	fsInfo fs.FsInfo,
+	cgroupSubsystems *containerlibcontainer.CgroupSubsystems,
+	inHostNamespace bool,
+	metadataEnvs []string,
+	ignoreMetrics container.MetricSet,
+) (container.ContainerHandler, error) {
+	// Create the cgroup paths.
+	cgroupPaths := make(map[string]string, len(cgroupSubsystems.MountPoints))
+	for key, val := range cgroupSubsystems.MountPoints {
+		cgroupPaths[key] = path.Join(val, name)
+	}
+
+	// Generate the equivalent cgroup manager for this container.
+	cgroupManager := &cgroupfs.Manager{
+		Cgroups: &libcontainerconfigs.Cgroup{
+			Name: name,
+		},
+		Paths: cgroupPaths,
+	}
+
+	id := ContainerNameToContainerdID(name)
+	// We assume that if load fails then the container is not known to containerd.
+	ctx := context.Background()
+	cntr, err := client.LoadContainer(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	var spec specs.Spec
+	if err := json.Unmarshal(cntr.Spec.Value, &spec); err != nil {
+		return nil, err
+	}
+
+	taskPid, err := client.TaskPid(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	rootfs := "/"
+	if !inHostNamespace {
+		rootfs = RootFSStorageDir + cntr.ID + spec.Root.Path
```

Comment:
This doesn't seem right to me. What rootfs do we need here? It seems that we are using this rootfs to get network stats.
However,
1) Why docker/cri-o are using `/rootfs`?
2) There may not be `proc` inside container, right? e.g. an image built from scratch.

Reply:
if its in non host namespace then proc is located at ```/var/run/containerd/io.containerd.runtime.v1.linux/k8s.io/<containerid>/rootfs```. This from what I verified during coding. Am I wrong here ?

Reply:
TBH, I don't know.
The pid we get is host pid, right? Why don't we just access `/proc/pid/net/ns`?

Reply:
I believe the `inHostNameSpace` here, means whether cadvisor is running in host namespace or inside container.

When cadvisor is running inside a container, we mount host root filesystem to `/rootfs`.

Reply:
So we should have same code with docker, and remove the container rootfs related logic.

Reply:
See [here](https://github.com/google/cadvisor/blob/master/manager/manager.go#L191-L196).